src/librbd/cache/ParentWriteLog.cc

```diff
@@ -0,0 +1,2769 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include <libpmemobj.h>
+#include "ParentWriteLog.h"
+#include "include/buffer.h"
+#include "include/Context.h"
+#include "include/ceph_assert.h"
+#include "common/deleter.h"
+#include "common/dout.h"
+#include "common/environment.h"
+#include "common/errno.h"
+#include "common/WorkQueue.h"
+#include "common/Timer.h"
+#include "common/perf_counters.h"
+#include "librbd/ImageCtx.h"
+#include "librbd/asio/ContextWQ.h"
+#include "librbd/cache/rwl/ImageCacheState.h"
+#include "librbd/cache/rwl/LogEntry.h"
+#include "librbd/cache/rwl/ReadRequest.h"
+#include "librbd/cache/rwl/Types.h"
+#include <map>
+#include <vector>
+
+#undef dout_subsys
+#define dout_subsys ceph_subsys_rbd_rwl
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::cache::ParentWriteLog: " << this << " " \
+                           <<  __func__ << ": "
+
+namespace librbd {
+namespace cache {
+
+using namespace librbd::cache::rwl;
+
+typedef ParentWriteLog<ImageCtx>::Extent Extent;
+typedef ParentWriteLog<ImageCtx>::Extents Extents;
+
+const unsigned long int OPS_APPENDED_TOGETHER = MAX_ALLOC_PER_TRANSACTION;
+
+template <typename I>
+ParentWriteLog<I>::ParentWriteLog(I &image_ctx, librbd::cache::rwl::ImageCacheState<I>* cache_state)
+  : m_cache_state(cache_state),
+    m_rwl_pool_layout_name(POBJ_LAYOUT_NAME(rbd_rwl)),
```

Comment:
Nit: the base class shouldn't have any PMDK-related functionality

Reply:
Yup. This PR is an initial framework that sets up upcoming code. In the next PR, a majority of PMDK-related code is moved to derived `ReplicatedWriteLog` class

Reply:
Currently all of `ReplicatedWriteLog` code is in `AbstractWriteLog` class. All pmdk functionality will be incrementally moved to `ReplicatedWriteLog`  after refactoring common code with SSD cache implementation